power_grid_model_c/power_grid_model/include/power_grid_model/grouped_index_vector.hpp

```diff
@@ -0,0 +1,255 @@
+                                                                    { ++t } -> std::same_as<T&>;
+                                                                    { --t } -> std::same_as<T&>;
+
+                                                                    { t + n } -> std::same_as<T>;
+                                                                    { t - n } -> std::same_as<T>;
+                                                                    { t += n } -> std::same_as<T&>;
+                                                                    { t -= n } -> std::same_as<T&>;
+                                                                };
+
+template <typename T, typename ElementType>
+concept random_access_iterable_like = requires(T const t) {
+                                          { t.begin() } -> random_access_iterator_like<ElementType>;
+                                          { t.end() } -> random_access_iterator_like<ElementType>;
+                                      };
+
+template <typename T>
+concept index_range_iterator =
+    random_access_iterator_like<T, typename T::iterator> && requires(T const t) {
+                                                                typename T::iterator;
+                                                                { *t } -> random_access_iterable_like<Idx>;
+                                                            };
+
+template <typename T>
+concept grouped_index_vector_type = std::default_initializable<T> && requires(T const t, Idx const idx) {
+                                                                         { t.size() } -> std::same_as<Idx>;
+
+                                                                         { t.begin() } -> index_range_iterator;
+                                                                         { t.end() } -> index_range_iterator;
+                                                                         {
+                                                                             t.get_element_range(idx)
+                                                                             } -> random_access_iterable_like<Idx>;
+
+                                                                         { t.element_size() } -> std::same_as<Idx>;
+                                                                         { t.get_group(idx) } -> std::same_as<Idx>;
+                                                                     };
+} // namespace detail
+
+template <typename T>
+concept grouped_idx_vector_type = detail::grouped_index_vector_type<T>;
+
+struct from_sparse_t {};
+struct from_dense_t {};
+
+constexpr auto from_sparse = from_sparse_t{};
+constexpr auto from_dense = from_dense_t{};
+
+class DenseGroupedIdxVector;
+
+class SparseGroupedIdxVector {
+  private:
+    class GroupIterator : public boost::iterator_facade<GroupIterator, Idx, boost::random_access_traversal_tag,
```

Comment:
I do not foresee the need of using random access iterator for the `GroupIterator`. A forward iterator should be enough. You only need to define `increment` function.